File: leetcode/001_two_sum/solution.rb
# frozen_string_literal: true

#
# Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

# You may assume that each input would have exactly one solution, and you may not use the same element twice.

# You can return the answer in any order.

# Example 1:

# Input: nums = [2,7,11,15], target = 9
# Output: [0,1]
# Explanation: Because nums[0] + nums[1] == 9, we return [0, 1].

# Example 2:

# Input: nums = [3,2,4], target = 6
# Output: [1,2]

# Example 3:

# Input: nums = [3,3], target = 6
# Output: [0,1]

# Constraints:

#     2 <= nums.length <= 104
#     -109 <= nums[i] <= 109
#     -109 <= target <= 109
#     Only one valid answer exists.

# Follow-up: Can you come up with an algorithm that is less than O(n2) time complexity?

# O(n)
# 3ms
# 212.44MB
def two_sum(nums, target)
  num_index = Hash.new(nil)
  nums.each_with_index do |num, index|
    complement = target - num
    return [num_index[complement], index] if num_index[complement]

    num_index[num] = index
  end
end

def also_slow_two_sum(nums, target)
  # Runtime: 399 ms - 20% off solutions
  # Memory Usage: 211 MB - 99% better than other submissions
  last_index = nums.size - 1
  nums.size.times do |pop_index|
    num = nums.pop # poping from tail is faster, shift wont pass time limit
    nums.each_with_index do |num2, index2|
      return [index2, last_index - pop_index] if (num + num2) == target
    end
  end
end

# n^2 solution
# Doesn't pass time limit on LEETCODE
def naive_two_sum(nums, target)
  nums.each_with_index do |num, index|
    nums[index.next..].each_with_index do |num2, index2|
      return [index, index.next + index2] if (num + num2) == target
    end
  end
end

RSpec.describe "Solution" do
  it "works" do
    expect(two_sum([2, 7, 11, 15], 9)).to eq([0, 1])
    expect(two_sum([3, 2, 4], 6)).to eq [1, 2]
    expect(two_sum([3, 3], 6)).to eq [0, 1]
  end
end

File: leetcode/001_two_sum/solution.py
import unittest
from typing import List


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # naive O(n^2)
        # Runtime316 ms Beats 42.6% Memory17.2 MB Beats 77.35%
        # size = len(nums)
        # for i in range(size):
        #     tail = nums.pop()
        #     for j in nums:
        #         if tail + j == target:
        #             return [nums.index(j), size - 1- i]

        # using HashMap O(n) for Time and Space
        # Runtime76 ms Beats 70.2% Memory17.7 MB Beats 35.80%
        hash_map = {}
        for i, arg_one in enumerate(nums):
            arg_two = target - arg_one
            if arg_two in hash_map:
                return [hash_map[arg_two], i]
            hash_map[arg_one] = i


s = Solution()
result = s.twoSum([2, 7, 11, 15], 9)
print(result)
